node/subsystem-util/src/reputation.rs

```diff
@@ -0,0 +1,117 @@
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! A utility abstraction to collect and send reputation changes.
+
+use polkadot_node_network_protocol::{PeerId, UnifiedReputationChange};
+use polkadot_node_subsystem::{
+	messages::{NetworkBridgeTxMessage, ReportPeerMessage},
+	overseer,
+};
+use std::{collections::HashMap, time::Duration};
+
+/// Default delay for sending reputation changes
+pub const REPUTATION_CHANGE_INTERVAL: Duration = Duration::from_secs(30);
+
+type BatchReputationChange = HashMap<PeerId, i32>;
+
+/// Collects reputation changes and sends them in one batch to relieve network channels
+#[derive(Debug, Clone)]
+pub struct ReputationAggregator {
+	send_immediately_if: fn(UnifiedReputationChange) -> bool,
+	by_peer: BatchReputationChange,
+}
+
+impl Default for ReputationAggregator {
+	fn default() -> Self {
+		Self::new(|rep| matches!(rep, UnifiedReputationChange::Malicious(_)))
+	}
+}
+
+impl ReputationAggregator {
+	/// New ReputationAggregator
+	///
+	/// # Arguments
+	///
+	/// * `send_immediately_if` - A function, takes `UnifiedReputationChange`,
+	/// results shows if we need to send the changes right away.
+	/// Useful for malicious changes and tests.
+	pub fn new(send_immediately_if: fn(UnifiedReputationChange) -> bool) -> Self {
+		Self { by_peer: Default::default(), send_immediately_if }
+	}
+
+	/// Sends collected reputation changes in a batch,
+	/// removing them from inner state
+	pub async fn send(
+		&mut self,
+		sender: &mut impl overseer::SubsystemSender<NetworkBridgeTxMessage>,
+	) {
+		if self.by_peer.is_empty() {
+			return
+		}
+
+		sender
+			.send_message(NetworkBridgeTxMessage::ReportPeer(ReportPeerMessage::Batch(
+				self.by_peer.clone(),
+			)))
+			.await;
+
+		self.by_peer.clear();
+	}
+
+	/// Adds reputation change to inner state
+	/// or sends it right away if the change is dangerous
+	pub async fn modify(
+		&mut self,
+		sender: &mut impl overseer::SubsystemSender<NetworkBridgeTxMessage>,
+		peer_id: PeerId,
+		rep: UnifiedReputationChange,
+	) {
+		if (self.send_immediately_if)(rep) {
+			self.single_send(sender, peer_id, rep).await;
+		} else {
+			self.add(peer_id, rep);
+		}
+	}
+
+	async fn single_send(
+		&self,
+		sender: &mut impl overseer::SubsystemSender<NetworkBridgeTxMessage>,
+		peer_id: PeerId,
+		rep: UnifiedReputationChange,
+	) {
+		sender
+			.send_message(NetworkBridgeTxMessage::ReportPeer(ReportPeerMessage::Single(
+				peer_id,
+				rep.into(),
+			)))
+			.await;
+	}
+
+	fn add(&mut self, peer_id: PeerId, rep: UnifiedReputationChange) {
+		add_reputation(&mut self.by_peer, peer_id, rep)
```

Comment:
I think that one thing that will be slightly different with this approach where we aggregate the reputation change will be here:`https://github.com/paritytech/substrate/blob/master/client/network/src/peer_store.rs#L161`.  However, since our flush interval is relatively small 30s, I think we should be fine.